/*
UNISTDX — C++ library for Linux system calls.
© 2018, 2019, 2020 Ivan Gankevich

This file is part of UNISTDX.

This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <http://unlicense.org/>
*/

#ifndef UNISTDX_BASE_BAD_CALL
#define UNISTDX_BASE_BAD_CALL

#include <iosfwd>
#include <system_error>

#include <unistdx/bits/macros>

namespace sys {

    /**
    \brief
    A \link std::system_error \endlink that records file, line and function where
    the exception was thrown.
    \date 2018-05-21
    */
    class bad_call: public std::system_error {

    private:
        const char* _file;
        const int _line;
        const char* _function;

    public:

        /// Construct the error with precise location where it was thrown.
        inline
        bad_call(
            std::errc errcond,
            const char* file,
            const int line,
            const char* function
        ) noexcept:
        std::system_error(int(errcond), std::generic_category()),
        _file(file), _line(line), _function(function) {}

        /// Construct the error with precise location where it was thrown.
        inline
        bad_call(
            const char* file,
            const int line,
            const char* function
        ) noexcept:
        bad_call(std::errc(errno), file, line, function) {}

        friend std::ostream&
        operator<<(std::ostream& out, const bad_call& rhs);

        /// Get error condition.
        inline std::errc
        errc() const noexcept {
            return std::errc(this->code().value());
        }

    };

    /// Output error message prefixed by the location where it was thrown.
    std::ostream&
    operator<<(std::ostream& out, const bad_call& rhs);

}

#endif // vim:filetype=cpp
